src/main/kotlin/com/sourcegraph/cody/ui/WebPanelProvider.kt

```diff
@@ -0,0 +1,110 @@
+package com.sourcegraph.cody.ui
+
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.fileEditor.*
+import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx
+import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.Key
+import com.intellij.openapi.vfs.VirtualFile
+import java.beans.PropertyChangeListener
+import javax.swing.JComponent
+import javax.swing.JLabel
+import org.jetbrains.annotations.NonNls
+
+class WebPanelEditor(private val file: VirtualFile) : FileEditor {
+  companion object {
+    val WEB_UI_PROXY_KEY = Key.create<WebUIProxy>("WebUIProxy")
+
+    var epoch = 0
+  }
+
+  val age = epoch++
+  private val userData: MutableMap<Any, Any?> = mutableMapOf()
```

Comment:
Is this guaranteed to be accesses from single thread only?

Reply:
Good feedback... JetBrains says we should delegate to a UserDataHolderBase. I read the implementation of it, it definitely handles multithreaded access.

Reply:
https://github.com/sourcegraph/jetbrains/pull/2085 for you.